edge-hub/watchdog/src/main.rs

```diff
@@ -0,0 +1,85 @@
+mod child;
+
+use std::{
+    io::Error,
+    sync::atomic::{AtomicBool, Ordering},
+    sync::Arc,
+    thread,
+};
+
+use anyhow::{Context, Result};
+use child::run;
+use signal_hook::{iterator::Signals, SIGINT, SIGTERM};
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+fn main() -> Result<()> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = register_shutdown_listener()
+        .context("Failed to register sigterm listener. Shutting down.")?;
+
+    let broker_handle = run(
+        "MQTT Broker".to_string(),
+        "/usr/local/bin/mqttd".to_string(),
+        "-c /tmp/mqtt/config/production.json".to_string(),
+        Arc::clone(&should_shutdown),
+    )?;
+
+    let edgehub_handle = match run(
+        "Edge Hub".to_string(),
+        "dotnet".to_string(),
+        "/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll".to_string(),
+        Arc::clone(&should_shutdown),
+    ) {
+        Ok(handle) => Some(handle),
+        Err(e) => {
+            should_shutdown.store(true, Ordering::Relaxed);
+            error!("Could not start edgehub process. {}", e);
+            None
+        }
+    };
+
+    if let Err(e) = broker_handle.join() {
+        should_shutdown.store(true, Ordering::Relaxed);
+        error!("Failure while running broker process. {:?}", e)
+    }
+    info!("Successfully stopped broker process");
+
+    edgehub_handle.map(|handle| {
+        if let Err(e) = handle.join() {
+            should_shutdown.store(true, Ordering::Relaxed);
+            error!("Failure while running edgehub process. {:?}", e);
+        }
+        info!("Successfully stopped edgehub process");
+    });
+
+    info!("Stopped watchdog process");
+    Ok(())
+}
+
+fn init_logging() {
+    let subscriber = Subscriber::builder().with_max_level(Level::INFO).finish();
+    let _ = subscriber::set_global_default(subscriber);
+}
+
+fn register_shutdown_listener() -> Result<Arc<AtomicBool>, Error> {
+    info!("Registering shutdown signal listener");
+    let shutdown_listener = Arc::new(AtomicBool::new(false));
+    let should_shutdown = shutdown_listener.clone();
+    let signals = Signals::new(&[SIGTERM, SIGINT])?;
+    thread::spawn(move || {
+        for signal in signals.forever() {
+            let signal_name = match signal {
+                SIGTERM => "SIGTERM",
+                SIGINT => "SIGINT",
+                _ => "Unexpected",
+            };
```

Comment:
I tried to map the int to a name automatically but didn't find a way to do that.

Reply:
I'd say you don't need to change  shutdown_listener into true for unexpected signals. What you can do it to apply https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map on the iterator
```rust
signals.forever()
.filter_map(|signal| match signal {
  SIGTERM => Some("SIGTERM"),
  SIGINT => Some("SIGINT"),
  _ => None})
.for_each(|signal| {
  info!("Watchdog received {} signal", signal);
  shutdown_listener.store(true, Ordering::Relaxed);
});
```